src/pages/Groups.tsx

```diff
@@ -68,147 +36,79 @@ export const Groups: MainStackScreen<"Groups"> = () => {
-            }
-        }
-    }, [prevSearch])
+        clearTimeout(searchTimeout)
+        searchTimeout = setTimeout(() => {
+            void searchGroups()
+        }, deltaTime)
+    }, [search])
 
     //if filters are applied after search, search again
     useEffect(() => {
-        if (isMounted && groups) {
-            forceSearch()
-        }
-    }, [course, year, type, platform])
-
-    //load groups to Ordered groups every time a new response from api arrives.
-    useEffect(() => {
-        if (isMounted) {
-            let newGroups = groups
-            //if response from api arrives when we no longer expect it
-            if (newGroups && search.length < 5) {
-                newGroups = undefined
-            }
-            if (year === all && newGroups) {
-                newGroups = orderByMostRecentYear(newGroups)
-            }
-            setOrderedGroups(newGroups)
-        }
-    }, [groups])
-
-    //force search in next render
-    const forceSearch = () => {
-        setRescheduleSearch(rescheduleSearch + 1)
-        setNeedSearching(true)
-    }
+        if (isMounted && groups) void searchGroups()
+    }, [filters])
 
     return (
         <PageWrapper>
             <View
                 style={{
                     flex: 1,
-                    marginHorizontal: 26,
+                    marginHorizontal: 28,
                     paddingTop: 56,
                 }}
             >
                 <Title>Gruppi Corsi</Title>
                 <AnimatedPoliSearchBar
-                    isSearching={isSearching}
-                    setIsSearching={val => setIsSearching(val)}
-                    setSearch={val => setSearch(val)}
-                    groups={orderedGroups}
+                    onSearch={val => setSearch(val)}
                     style={{ marginTop: 36, marginBottom: 22 }}
                 />
                 <Filters
-                    year={year}
-                    setYear={val => setYear(val)}
-                    course={course}
-                    setCourse={val => setCourse(val)}
-                    type={type}
-                    setType={val => setType(val)}
-                    platform={platform}
-                    setPlatform={val => setPlatform(val)}
-                    forceSearch={forceSearch}
+                    onFilterChange={filters => setFilters(filters)}
+                    filters={filters}
                 />
 
                 <View
                     style={{
                         flex: 1,
                         marginTop: 40,
                         marginBottom: 93,
+                        marginHorizontal: 8,
                     }}
                 >
                     <ScrollView>
-                        {orderedGroups?.map((group, idx) => {
+                        {orderByMostRecentYear(
+                            groups,
+                            filters.year !== undefined
+                        )?.map((group, idx) => {
```

Comment:
sai cosa ahah, avevi ragione te effettivamente, non ci avevo fatto caso a questa riga ma chiamare la funzione inline in questo modo effettivamente è un po' bruttino, soprattuto perché hai dovuto aggiungere quel secondo parametro, però niente ti impedisce di usare una normalissima variabile senza toccare lo state appena prima del return statement

```tsx
const orderedGroups = filters.year === undefined ? orderByMostRecentYear(groups) : groups

return (
    <PageWrapper>
        ...
        <ScrollView>
            {orderedGroups.map((group, idx) => {
            ...